baseapp/abci.go

```diff
@@ -343,6 +343,36 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
 	// The write to the DeliverTx state writes all state transitions to the root
 	// MultiStore (app.cms) so when Commit() is called is persists those values.
 	app.deliverState.ms.Write()
+
+	// Check if there has been an app version change. If so and there is a migrator
+	// set, begin to run migrations. This needs to be done before the commit so
+	// that the migrations are part of the app hash
+	if header.Version.App < app.appVersion &&
+		app.migrator.storeMigrator != nil &&
+		app.migrator.moduleMigrator != nil {
+
+		// first update the stores themselves by adding and removing them as necessary
+		storeMigrations, err := app.migrator.storeMigrator(header.Version.App, app.appVersion)
+		if err != nil {
+			panic(fmt.Sprintf("failed to get store migrations: %v", err))
+		}
+		app.MountKVStores(storeMigrations.Added)
+		err = app.cms.LoadLatestVersionAndUpgrade(storeMigrations.ToStoreUpgrades())
+		if err != nil {
+			panic(fmt.Sprintf("failed to upgrade stores: %v", err))
+		}
+
+		// create a new cached branch of the store to apply migrations to
+		app.setDeliverState(header)
+		err = app.migrator.moduleMigrator(app.deliverState.ctx, header.Version.App, app.appVersion)
+		if err != nil {
+			panic(fmt.Sprintf("failed to migrate modules: %v", err))
+		}
```

Comment:
We need to be careful about atomicity here. I need to double check how CometBFT would handle this panic but iirc, we should also revert all the transactions previously committed